cmake/get_cpm.cmake

```diff
@@ -15,16 +16,23 @@ function(download_cpm)
   message(STATUS "Downloading CPM.cmake to ${CPM_DOWNLOAD_LOCATION}")
   file(DOWNLOAD
        https://github.com/cpm-cmake/CPM.cmake/releases/download/v${CPM_DOWNLOAD_VERSION}/CPM.cmake
-       ${CPM_DOWNLOAD_LOCATION}
+       ${CPM_DOWNLOAD_LOCATION} STATUS CPM_DOWNLOAD_STATUS
   )
+  list(GET CPM_DOWNLOAD_STATUS 0 CPM_DOWNLOAD_STATUS_CODE)
+  list(GET CPM_DOWNLOAD_STATUS 1 CPM_DOWNLOAD_ERROR_MESSAGE)
+  if(${CPM_DOWNLOAD_STATUS_CODE} EQUAL 0)
+    message(STATUS "CPM: Download completed successfully.")
+  else()
+    message(FATAL_ERROR "CPM: Error occurred during download: ${CPM_DOWNLOAD_ERROR_MESSAGE}")
+  endif()
 endfunction()
 
 if(NOT (EXISTS ${CPM_DOWNLOAD_LOCATION}))
   download_cpm()
 else()
   # resume download if it previously failed
-  file(READ ${CPM_DOWNLOAD_LOCATION} check)
-  if("${check}" STREQUAL "")
+  file(SHA256 ${CPM_DOWNLOAD_LOCATION} CPM_CHECK)
+  if(NOT "${CPM_CHECK}" STREQUAL CPM_HASH_SUM)
```

Comment:
I've added both checks, since I think they are both useful. You're right, the straightforward `EXPECTED_HASH` variable within `file()` is the best solution for the download - but I also want to have a check for any existing file.

Reply:
> but I also want to have a check for any existing file

From the CMake doc above:

> If the file already exists and matches the hash, the download is skipped.

As I understand it, this also check existing files and re-downloads if the hash doesn't match the expectation.

The docs also state the following:

> If after download the file does not match the hash, the operation fails with an error.

So if the operation completes without error, I think we can safely assume that the download has finished successfully and CPM is integrity checked.

So as I understand it means we should be able to replace most in this file following line 15 with a single `file(DOWNLOAD ...)` call followed by `include(${CPM_DOWNLOAD_LOCATION})`? This would be pretty neat as it reduces boilerplate code used by most projects that use CPM.

Reply:
But doesn't this also mean that we could get rid of the entire `download_cpm` function? This function is then only called once (since we won't have to differentiate between existing and non-existing `${CPM_DOWNLOAD_LOCATION}`s), so now we could write this part without the surrounding `function`…

Reply:
Agree 100%

Reply:
Reworked the `cmake/get_cpm.cmake` accordingly.